### PR TITLE
Queue for SwaV

### PR DESCRIPTION
**Pull request by ibro45**
Aims to solve https://github.com/lightly-ai/lightly/issues/1006

Implements queue for SwaV with minimal code change in the training loop. Implemented as per details from the paper:

Rationale

image

A trick for small batch sizes - start using queue prototypes at a later epoch

image

### CODE REVIEW

File: lightly/loss/swav_loss.py [hunk 1]
```diff
@@ -110,7 +110,8 @@ def subloss(self, z: torch.Tensor, q: torch.Tensor):
 
     def forward(self,
                 high_resolution_outputs: List[torch.Tensor],
-                low_resolution_outputs: List[torch.Tensor]):
+                low_resolution_outputs: List[torch.Tensor],
+                queue_outputs: List[torch.Tensor]):
```
Comment (guarin) [inline lightly/loss/swav_loss.py:114]: Could you make the `queue_outputs` optional (default is `None`)?

File: lightly/models/modules/heads.py [hunk 2]
```diff
@@ -371,7 +371,7 @@ class SwaVPrototypes(nn.Module):
     """Multihead Prototypes used for SwaV.
 
     Each output feature is assigned to a prototype, SwaV solves the swapped
-    predicition problem where the features of one augmentation are used to
+    prediction problem where the features of one augmentation are used to
```
Comment (guarin) [inline lightly/models/modules/heads.py:374]: Thanks for fixing this one :)

File: lightly/models/swav.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class SwaV(nn.Module):
+    def __init__(self, backbone, num_ftrs, out_dim, n_prototypes, queue_length=0):
+        super().__init__()
+        self.backbone = backbone
+        self.projection_head = SwaVProjectionHead(num_ftrs, num_ftrs, out_dim)
+        self.prototypes = SwaVPrototypes(out_dim, n_prototypes=n_prototypes)
+        # Queues are initialized in the first forward call
+        self.queues = []
```
Comment (guarin) [inline lightly/models/swav.py:15]: I think you could register the queues in the beginning as a [ModuleList](https://pytorch.org/docs/stable/generated/torch.nn.ModuleList.html):

```
self.queues = ModuleList(modules=[MemoryBankModule(size=queue_length) for _ in range(n_queues)])
```

This would make sure that queues are moved to the correct device on `model.to(device)` calls. The drawback is that `n_queues` has to be passed as argument but I think that it is ok to assume that the number of queues is known from the beginning.

What do you think?

File: lightly/models/swav.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import torch
+from torch import nn
+
+from lightly.models.modules import SwaVProjectionHead, SwaVPrototypes
+from lightly.loss.memory_bank import MemoryBankModule
```
Comment (guarin) [inline lightly/models/swav.py:5]: We decided a while ago to not add "full" models to `lightly/models` anymore. The reason was that users usually want to customize models in some way or another and would have to change the source code to do so. Instead we started writing example model implementations in `examples/` that are easy to copy paste and modify.

Could you move everything in this file to `swav_new.py` (maybe rename the file to `swav_queue.py` or similar)? Then it would perfectly follow our examples "philosophy".

File: lightly/models/swav.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        self.prototypes.normalize()
+
+        high_resolution_features = [self._subforward(x) for x in high_resolution]
+        low_resolution_features = [self._subforward(x) for x in low_resolution]
+        queue_features = self._get_queue_features(high_resolution_features)
+
+        high_resolution_prototypes = [self.prototypes(x) for x in high_resolution_features]
+        low_resolution_prototypes = [self.prototypes(x) for x in low_resolution_features]
+        queue_prototypes = [self.prototypes(x) for x in queue_features]
+
```
Comment (guarin) [inline lightly/models/swav.py:30]: I believe this line should be in a `torch.no_grad()` block:
```
queue_prototypes = [self.prototypes(x) for x in queue_features]
```

The original repo does queue prototype assignment also with `torch.no_grad()`: https://github.com/facebookresearch/swav/blob/06b1b7cbaf6ba2a792300d79c7299db98b93b7f9/main_swav.py#L291-L301

**Review by guarin — COMMENTED** [review]
Hi, thanks a ton for the lightning fast PR!

I had a look at the code and this looks very promising. Let me know if I can somehow help you with the memory bank issue. Are you aware that there was a very recent change to the memory bank in #1009 ?

**Comment by ibro45** [comment]
Thanks so much for your feedback! And yes, I did fork from the last commit with the MemoryBank fix

**Comment by guarin** [comment]
Awesome! Were you able to fix the issue with the memory bank? 
Let me know if the PR is ready for review again.

**Comment by ibro45** [comment]
This was the issue with the memory bank. https://github.com/lightly-ai/lightly/blob/cf30f901b942eda7c426ad2b456b1f5d75504a91/examples/pytorch/swav_queue.py#L58

Is there a reason why the memory bank output dimensions are swapped?

**Comment by guarin** [comment]
> Is there a reason why the memory bank output dimensions are swapped?

I am not aware of any particular reason. I guess we followed to original implementation where it is also swapped: https://github.com/facebookresearch/moco/blob/78b69cafae80bc74cd1a89ac3fb365dc20d157d3/moco/builder.py#L39

**Comment by ibro45** [comment]
I see, then I'll remove those todo comments and leave as is. Otherwise it's ready for a review

**Review by guarin — APPROVED** [review]
This looks great, thanks a lot for the PR!

We'll try to run the code on one of our benchmarks :)

**Comment by ibro45** [comment]
Awesome, thank you :)

**Comment by guarin** [comment]
Hi @ibro45, I tested the queue implementation and it seems to work (accuracy >45% on Imagenet100 after 50 epochs with batch size 128 on a Tesla T4). The accuracy only seems to increase after epoch 15 once the queue is used. Before, the accuracy was always hovering around 1%. Maybe one could use the queue even earlier to speed up training, although the paper mentions that the queue should not be used too early.

To verify that the implementation matches the original we would have to train on full Imagenet but we don't have time to do that right now. I pushed the code I used here #1015 if you want to have a look (had to adapt it to pytorch lightning).

**Comment by ibro45** [comment]
That's great to hear!

I think you're right, while 15 epochs before the queue were good for ImageNet, maybe a lower number would be more suitable for ImageNet100.

On another note, do you think it'd be possible to add the plots of losses and metrics of these models? A Weights&Biases project that includes all the models would be amazing for newbies like me to know how these models behave. For example, someone training a SwaV might get alarmed about the accuracy being that low before the queue kicks. Actually, that's what I'm struggling with right now, I'm trying to apply it to medical images and it's difficult to know what's normal and what needs more parameter tuning.
